ArduCopter/mode_auto.cpp

```diff
@@ -1165,21 +1167,191 @@ void ModeAuto::payload_place_run()
+            // do nothing on this loop
+            break;
+        case PayloadPlaceStateType_Descent:
+            gcs().send_text(MAV_SEVERITY_INFO, "%s landed", prefix_str);
+            nav_payload_place.state = PayloadPlaceStateType_Release;
+            break;
+        case PayloadPlaceStateType_Release:
+        case PayloadPlaceStateType_Releasing:
+        case PayloadPlaceStateType_Delay:
+        case PayloadPlaceStateType_Ascent_Start:
+        case PayloadPlaceStateType_Ascent:
+        case PayloadPlaceStateType_Done:
+            break;
+        }
+    }
+
+#if AP_GRIPPER_ENABLED == ENABLED
+    // if pilot releases load manually:
+    if (g2.gripper.valid() && g2.gripper.released()) {
+        switch (nav_payload_place.state) {
+        case PayloadPlaceStateType_FlyToLocation:
+        case PayloadPlaceStateType_Descent_Start:
+            set_submode(SubMode::NAV_PAYLOAD_PLACE);
+            gcs().send_text(MAV_SEVERITY_INFO, "%s Manual release", prefix_str);
+            nav_payload_place.state = PayloadPlaceStateType_Done;
+            break;
+        case PayloadPlaceStateType_Descent:
+            gcs().send_text(MAV_SEVERITY_INFO, "%s Manual release", prefix_str);
+            nav_payload_place.state = PayloadPlaceStateType_Release;
+            break;
+        case PayloadPlaceStateType_Release:
+        case PayloadPlaceStateType_Releasing:
+        case PayloadPlaceStateType_Delay:
+        case PayloadPlaceStateType_Ascent_Start:
+        case PayloadPlaceStateType_Ascent:
+        case PayloadPlaceStateType_Done:
+            break;
+        }
+    }
+#endif
+
+    switch (nav_payload_place.state) {
+    case PayloadPlaceStateType_FlyToLocation:
+        if (copter.wp_nav->reached_wp_destination()) {
+            payload_place_start();
+        }
+        break;
+
+    case PayloadPlaceStateType_Descent_Start:
+        nav_payload_place.descent_established_time_ms = now_ms;
+        nav_payload_place.descent_start_altitude_cm = inertial_nav.get_position_z_up_cm();
+        nav_payload_place.descent_speed_cms = (is_positive(g2.pldp_descent_speed_ms)) ? -g2.pldp_descent_speed_ms * 100.0 : -abs(g.land_speed);
+        nav_payload_place.descent_thrust_level = 1.0;
+        nav_payload_place.state = PayloadPlaceStateType_Descent;
+        FALLTHROUGH;
+
+    case PayloadPlaceStateType_Descent:
+        // check maximum decent distance
+        if (!is_zero(nav_payload_place.descent_max_cm) &&
+            nav_payload_place.descent_start_altitude_cm - inertial_nav.get_position_z_up_cm() > nav_payload_place.descent_max_cm) {
+            nav_payload_place.state = PayloadPlaceStateType_Ascent_Start;
+            gcs().send_text(MAV_SEVERITY_WARNING, "%s Reached maximum descent", prefix_str);
+            break;
+        }
+        // calibrate the decent thrust after aircraft has reached constant decent rate and release if threshold is reached
+        if (pos_control->get_vel_desired_cms().z > -0.95 * abs(g.land_speed)) {
```

Comment:
probably need to use nav_payload_place.descent_speed_cms here instead of land_speed?

Reply:
Good catch!!